linkup-cli/src/services/cloudflare_tunnel/mod.rs

```diff
@@ -0,0 +1,363 @@
+    }
+}
+
+impl BackgroundService<Error> for CloudflareTunnel {
+    const NAME: &str = "Cloudflare Tunnel";
+
+    async fn run_with_progress(
+        &self,
+        state: &mut LocalState,
+        status_sender: std::sync::mpsc::Sender<super::RunUpdate>,
+    ) -> Result<(), Error> {
+        if !state.should_use_tunnel() {
+            self.notify_update_with_details(
+                &status_sender,
+                super::RunStatus::Skipped,
+                "Requested no tunnel",
+            );
+        }
+
+        if signal::get_running_pid(&self.pidfile_path).is_some() {
+            self.notify_update_with_details(
+                &status_sender,
+                super::RunStatus::Started,
+                "Was already running",
+            );
+
+            return Ok(());
+        }
+
+        if Self::use_paid_tunnels() {
+            self.notify_update_with_details(&status_sender, super::RunStatus::Starting, "Paid");
+
+            if let Err(e) = self.start_paid().await {
+                self.notify_update_with_details(
+                    &status_sender,
+                    super::RunStatus::Error,
+                    "Failed to start",
+                );
+
+                return Err(e);
+            }
+        } else {
+            self.notify_update_with_details(&status_sender, super::RunStatus::Starting, "Free");
+
+            if let Err(e) = self.start_free() {
+                self.notify_update_with_details(
+                    &status_sender,
+                    super::RunStatus::Error,
+                    "Failed to start",
+                );
+
+                return Err(e);
+            }
+        }
+
+        // Pidfile existence check
+        {
+            let mut pid_file_ready_attempt = 0;
+            let mut pid_file_exists = signal::get_running_pid(&self.pidfile_path).is_some();
+            while !pid_file_exists && pid_file_ready_attempt <= 10 {
+                sleep(Duration::from_secs(1)).await;
+                pid_file_ready_attempt += 1;
+
+                self.notify_update_with_details(
+                    &status_sender,
+                    super::RunStatus::Starting,
+                    format!("Waiting for tunnel... retry #{}", pid_file_ready_attempt),
+                );
+
+                pid_file_exists = signal::get_running_pid(&self.pidfile_path).is_some();
+            }
+
+            if !pid_file_exists {
+                self.notify_update_with_details(
+                    &status_sender,
+                    super::RunStatus::Error,
+                    "Failed to start tunnel",
+                );
+
+                return Err(Error::PidfileNotFound);
+            }
+
+            self.notify_update(&status_sender, super::RunStatus::Starting);
+        }
+
+        // DNS Propagation check
+        {
+            let mut dns_propagation_attempt = 0;
+            let mut dns_propagated = self.dns_propagated().await;
+            // TODO: Isn't 40 too much?
```

Comment:
The free tunnels can be crazy slow - although it might be better to just have 20 and then fail

Reply:
How valid is the DNS propagation check in the paid case? Maybe it still is but we expect it to be faster?

Reply:
> The free tunnels can be crazy slow - although it might be better to just have 20 and then fail

Changed in c297e20e1388db88f214c8a772e0bdd6022d852b

Reply:
> How valid is the DNS propagation check in the paid case? Maybe it still is but we expect it to be faster?

Not sure if is valid 🤔 But it basically skip it because is super fast and it just works, so I think it doesn't hurt? 🤔

Reply:
Yeah sure leave it!